pyop2/types/dat.py

```diff
@@ -592,11 +593,58 @@ def local_to_global_end(self, insert_mode):
 
         :kwarg insert_mode: insertion mode (an access descriptor)"""
         halo = self.dataset.halo
-        if halo is None:
+        if halo is None or self._halo_frozen:
             return
         halo.local_to_global_end(self, insert_mode)
         self.halo_valid = False
 
+    @contextlib.contextmanager
+    @mpi.collective
+    def frozen_halo(self, access_mode):
+        """Temporarily disable halo exchanges inside a context manager.
+
+        :arg access_mode: Mode with which the data will subsequently be accessed.
+
+        This is useful in cases where one is repeatedly writing to a :class:`Dat` with
+        the same access descriptor since the intermediate updates can be skipped.
+        """
+        if access_mode in {Access.READ, Access.RW}:
+            raise ValueError(
+                "Trying to read from a Dat with a frozen halo will lead to unexpected "
+                "results. Freezing halos is only valid for write-only access modes."
+            )
+
+        # First initialise the halo values (e.g. set to zero if INC'ing)
+        self.global_to_local_begin(access_mode)
+        self.global_to_local_end(access_mode)
+        self.freeze_halo()
+
+        # Execute main body of context manager
+        yield self
```

Comment:
Another reason not to use `contextlib.contextmanager`: you _still_ need to manage exception handling in the body of your function. If executing the body raises an error, you won't run the cleanup code, which is kind of unexpected:

```python
In [7]: @contextlib.contextmanager
   ...: def foo():
   ...:     print("before")
   ...:     yield
   ...:     print("after")
   ...: 

In [8]: with foo():
   ...:     print("body")
   ...: 
before
body
after

In [9]: with foo():
   ...:     print("body")
   ...:     raise RuntimeError
   ...: 
before
body
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [9], in <cell line: 1>()
      1 with foo():
      2     print("body")
----> 3     raise RuntimeError

RuntimeError: 
```

Reply:
No, just remember to do it if you want the cleanup to occur!

Reply:
So:
```python
try:
    yield self
finally:
    self.unfreeze_halo()
    self.local_to_global(..)
    ...
```

Reply:
Sure. I was confused because I don't think of halo exchanges as 'cleanup' necessarily.

Reply:
I guess if something fails in the parloop body there's probably not a lot you can end up doing, but maybe you could imagine catching some things and you would want a consistent state for your halo afterwards.